### PR TITLE
feat: Add support for rust-rocksdb storage engine

### PR DESCRIPTION
**Pull request by creatorrr**
- **wip: Add support for rust-rocksdb storage engine**

https://crates.io/crates/rocksdb

### CODE REVIEW

**Comment by creatorrr** [comment]
Comments / thoughts @zh217 ?

**Comment by keitharobertson** [comment]
I'm doing some testing and this is working for me and even fixed a bug with the rocksdb windows thread shutdown

**Comment by creatorrr** [comment]
@keitharobertson I’m not sure why the tests are failing though. Haven’t had a chance to sit and debug. Maybe I missed something?

File: cozo-core/src/storage/newrocks.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+                    {
+                        extend_tuple_from_v(&mut tup, &v);
+                        return Some(Ok(tup));
+                    }
+                }
+                Err(e) => return Some(Err(miette!("Iterator error: {}", e))),
+            }
+        }
+        None
+    }
```
Comment (keitharobertson) [inline cozo-core/src/storage/newrocks.rs:379]:
```suggestion
    fn next(&mut self) -> Option<Self::Item> {
        loop {
            self.inner.set_mode(rocksdb::IteratorMode::From(&self.next_bound, rocksdb::Direction::Forward));
            match self.inner.next() {
                None => return None,
                Some(Ok((k_slice, v_slice))) => {
                    if self.upper_bound.as_slice() <= k_slice.as_ref() {
                        return None;
                    }

                    let (ret, nxt_bound) = check_key_for_validity(k_slice.as_ref(), self.valid_at, None);
                    self.next_bound = nxt_bound;
                    if let Some(mut tup) = ret {
                        extend_tuple_from_v(&mut tup, v_slice.as_ref());
                        return Some(Ok(tup));
                    }
                },
                Some(Err(e)) => return Some(Err(miette!("Iterator Error: {}", e))),
            }
        }
    }
```
Modeled off the implementation in rocks.rs, this solves your time travel test issues.

File: cozo-core/src/storage/newrocks.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+            Default::default(),
+            ScriptMutability::Immutable,
+        )?;
+        assert_eq!(result.rows[0][0], DataValue::from(200));
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_range_operations() -> Result<()> {
```
Comment (keitharobertson) [inline cozo-core/src/storage/newrocks.rs:505]: This test does not work for the old rocks implementation either. The problem is it is attempting to seek for the key with values (appended to the keys) included in the encoded prefix, which causes it to skip the first result. I can put a PR up with a fix, but it is unrelated to your implementation here.

**Review by keitharobertson — COMMENTED** [review]
Nice job finding bugs with the tests

**Comment by creatorrr** [comment]
Thanks @keitharobertson! This is awesome. Committed your suggested changes. Another question is naming? `storage-new-rocksdb` and new-* seem weird to me… Maybe we can switch to rocksdb for the new one and rename the original to cozorocks-* ?

@zh217 what do you think?

File: cozo-core/src/storage/newrocks.rs (hunk 2 above)
Comment (creatorrr) [reply cozo-core/src/storage/newrocks.rs:507]: Gotcha

**Comment by creatorrr** [comment]
It won't be a breaking change at the cozoscript level though, right? I like rocksdb-9 naming scheme though.

**Comment by keitharobertson** [comment]
> It won't be a breaking change at the cozoscript level though, right? I like rocksdb-9 naming scheme though.

yeah- I think it would only matter in Cargo.toml

**Comment by creatorrr** [comment]
> yeah- I think it would only matter in Cargo.toml 

Actually I think in `cozo-bin`, we might have to add a new flag too

**Pull request merged**